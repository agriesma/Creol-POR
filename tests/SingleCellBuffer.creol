/* SingleCellBuffer.creol

   by Einar Broch Johnson */

interface Buff
begin
op put(in x : Data)
op get(out x : Data)
end

class Buffer implements Buff
begin
var cell : Data
var ctr : Int
var next : Buff

op init == ctr := 0;
op run == skip;
op put(in x : Data) ==
    if ctr = 0 then
      cell := x ;
    else
      if next = null then
	next := new Buffer();
      fi
      next.put(x;);
    fi
    ctr := ctr + 1;

op get(out x : Data) ==
    await (ctr > 0) ; 
    if cell /= null then
      x := cell;
    else
      next.get(; x);
    fi
    ctr := ctr - 1;
    cell := null;

end
