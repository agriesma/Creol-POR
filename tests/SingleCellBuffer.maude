load interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Buffer" : Cl | Inh: noInh, Par: noAid, Att: "cell" |-> null , "ctr" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctr" ::= int(0) ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "put" : Mtdname | Param: "x", Latt: "x" |-> null, Code: if "=" ( "ctr" # int(0) ) th "cell" ::= "x" el if "=" ( "next" # null ) th "next" ::= new "Buffer" ( emp ) el skip fi ; "sync.0" ! "next" . "put" ( "x" ) ; ( "sync.0" ? ( noAid ) )  fi ; "ctr" ::= "+" ( "ctr" # int(1) ) ; return ( emp ) > *
  < "get" : Mtdname | Param: noAid, Latt: "x" |-> null, Code: ( await ">" ( "ctr" # int(0) ) ) ; if "/=" ( "cell" # null ) th "x" ::= "cell" el "sync.0" ! "next" . "get" ( emp ) ; ( "sync.0" ? ( "x" ) )  fi ; "ctr" ::= "-" ( "ctr" # int(1) ) ; "cell" ::= null ; return ( "x" ) >, Ocnt: 0 >

.
endm
