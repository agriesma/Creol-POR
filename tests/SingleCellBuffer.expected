==========================================
reduce in PROGRAM : init .
result Class: 
< 'Buffer : Cl | Inh: noInh,Par: noAid,Att: 'cell |-> null, 'ctr |-> null,
    'next |-> null,Mtds:
< 'get : Mtdname | Param: noAid,Latt: 'x |-> null,Code: await ('less[[int(0) #
    'ctr]]) ; if 'not[['equal[['cell # null]]]] th 'x ::= 'cell el 'next .
    'get(emp : 'x) fi ; 'ctr ::= ('minus[['ctr # int(1)]]) ; ('cell assign
    null) ; return('x) > *
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: ('ctr assign int(0)) ;
    return(emp) > *
< 'put : Mtdname | Param: 'x,Latt: 'x |-> null,Code: if 'equal[['ctr # int(0)]]
    th 'cell ::= 'x el if 'equal[['next # null]] th 'next ::= new 'Buffer(emp)
    el skip fi ; 'next . 'put('x : noAid) fi ; 'ctr ::= ('plus[['ctr # int(
    1)]]) ; return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: skip ; return(emp) >,Ocnt:
    0 >

Bye.
